Assignment4/PointGrid.py

```diff
@@ -0,0 +1,170 @@
+
+    def find_point(self, point_value):
+        """
+        Return the first NOT VISITED Point in self._points with the given value if exists. None otherwise.
+        :param point_value: any
+        :return: Point or None
+        """
+        for coord in self._points:
+            point = self.get_point(coord)
+            if point.value == point_value and point.is_visited == False:
+                return point
+
+    def get_point(self, coordinates):
+        """
+        Get element in Grid by its coordinates. Raises ValueError if coordinates are invalid.
+        :param coordinates: tuple(int,int)
+        :return: Point or raise ValueError
+        """
+        if not self.valid_coordinates(coordinates):
+            raise ValueError("invalid coordinates")
+        return self._points[coordinates]
+
+    def get_element(self, coordinates):
+        """
+        Return element in the Grid by its coordinates.
+        :param coordinates: tuple(int, int)
+        :return: any
+        """
+        return self._grid[coordinates[0]][coordinates[1]]
+
+    def valid_coordinates(self, coordinates):
+        """
+        Check if the coordinates are valid. Return True if so, False otherwise.
+        :param coordinates: tuple(int, int)
+        :return: bool
+        """
+        Rs, Cs = self.size()
+        return coordinates[0] in range(0, Rs) and coordinates[1] in range(0, Cs)
+
+    def _find_neighbor_land(self, point):
+        """
+        Returns a list of all neighbors of the point that are 'land'.
+        :param point: Point
+        :return: list(Points)
+        """
+        Rs, Cs = self.size()
+        r, c = point.coordinates[0], point.coordinates[1]
+
+        possible_coordinates = [(r-1, c), (r, c-1), (r, c+1), (r+1, c)]
+        return [self.get_point(coo) for coo in possible_coordinates
+                if (self.valid_coordinates(coo) and self.is_land(coo))]
+
+    def is_land(self, coordinates):
+        """
+        Return True if cell with such a coordinates is land - the value is 'T', False otherwise.
+        :param coordinates: tuple(int, int)
+        :return: bool
+        """
+        return self.get_element(coordinates) == 'T'
+
+    def find_land(self):
+        """
+        Find all the points that have value 'T' and return them in a list.
+        :return: list[Points]
+        """
+        land = [self.get_point(coo) for coo in self._points if self.is_land(coo)]
+        return land if land else None
+
+    ################################# FIND ISLANDS HERE ###########################################
+
+    '''
+    We find islands as lists of coordinates just to make it easier to test and to print them out.
+    If you find it better to have them as list of Points just replace "island.append(point.coordinates)"
+    in _recursive() with "island.append(point)".
+    '''
+
+    def find_islands(self):
+        """
+        Find islands in a grid and return a list of lists of points

```

Comment:
It looks like this method can only be called once because `is_visited` will be updated at runtime and thus the second call would probably return `None`? May be worth adding a test case with two calls to find_islands and ensuring they both return the correct value.